### PR TITLE
Handle h11.Connection.next_event() RemoteProtocolException

### LINKED ISSUE #96: h11._util.RemoteProtocolError: can't handle event type ConnectionClosed when role=SERVER and state=SEND_RESPONSE
I intermittently got this error when load testing uvicorn endpoint.

This error comes from a proxy endpoint where I am also using `encode/http3` to perform HTTP client calls.

```
  File "/project/venv/lib/python3.7/site-packages/http3/client.py", line 365, in post
    timeout=timeout,
  File "/project/venv/lib/python3.7/site-packages/http3/client.py", line 497, in request
    timeout=timeout,
  File "/project/venv/lib/python3.7/site-packages/http3/client.py", line 112, in send
    allow_redirects=allow_redirects,
  File "/project/venv/lib/python3.7/site-packages/http3/client.py", line 145, in send_handling_redirects
    request, verify=verify, cert=cert, timeout=timeout
  File "/project/venv/lib/python3.7/site-packages/http3/dispatch/connection_pool.py", line 121, in send
    raise exc
  File "/project/venv/lib/python3.7/site-packages/http3/dispatch/connection_pool.py", line 116, in send
    request, verify=verify, cert=cert, timeout=timeout
  File "/project/venv/lib/python3.7/site-packages/http3/dispatch/connection.py", line 59, in send
    response = await self.h11_connection.send(request, timeout=timeout)
  File "/project/venv/lib/python3.7/site-packages/http3/dispatch/http11.py", line 65, in send
    event = await self._receive_event(timeout)
  File "/project/venv/lib/python3.7/site-packages/http3/dispatch/http11.py", line 109, in _receive_event
    event = self.h11_state.next_event()
  File "/project/venv/lib/python3.7/site-packages/h11/_connection.py", line 439, in next_event
    exc._reraise_as_remote_protocol_error()
  File "/project/venv/lib/python3.7/site-packages/h11/_util.py", line 72, in _reraise_as_remote_protocol_error
    raise self
  File "/project/venv/lib/python3.7/site-packages/h11/_connection.py", line 422, in next_event
    self._process_event(self.their_role, event)
  File "/project/venv/lib/python3.7/site-packages/h11/_connection.py", line 238, in _process_event
    self._cstate.process_event(role, type(event), server_switch_event)
  File "/project/venv/lib/python3.7/site-packages/h11/_state.py", line 238, in process_event
    self._fire_event_triggered_transitions(role, event_type)
  File "/project/venv/lib/python3.7/site-packages/h11/_state.py", line 253, in _fire_event_triggered_transitions
    .format(event_type.__name__, role, self.states[role]))
h11._util.RemoteProtocolError: can't handle event type ConnectionClosed when role=SERVER and state=SEND_RESPONSE
```

### PR DESCRIPTION
**Pull request by toppk**
here is a proposed fix.  I wasn't too clear on the test suite (as usual :), but I was able to make a case that triggers the failure mode and test that the fix works as expected.

Addresses github issue #96

Fixes #96

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -103,6 +105,22 @@ def backend(request):
     await send({"type": "http.response.body", "body": b"Hello, world!"})
 
 
+async def server_crash(scope, receive, send):
```
Comment (florimondmanca) [inline tests/conftest.py]: Just a suggestion on naming: isn't this endpoint really more of a `server_premature_close`?

File: tests/conftest.py [hunk 2]
```diff
@@ -103,6 +105,22 @@ def backend(request):
...
+    crash_ms_str: str = scope["path"].replace("/server_crash/", "")
+    try:
+        crash_ms = float(crash_ms_str)
+    except ValueError:
+        crash_ms = 100
+    await asyncio.sleep(crash_ms / 1000.0)
```
Comment (florimondmanca) [inline tests/conftest.py]: Is it important to wait for a bit before closing the connection prematurely? (Just wondering if we can simplify the failing test case even further)

File: tests/conftest.py (hunk 1 above)
Comment (toppk) [reply tests/conftest.py]: premature close is more precise, I was naming it after what would typically cause a premature.  I can rename it.

**Comment by toppk** [comment]
> Thanks for this! Overall and after reading through the digging in #96, this looks sensible to me.
> 
> Some notes:
> 
>     * I'm not able to reproduce the results I got in [#96 (comment)](https://github.com/encode/httpx/issues/96#issuecomment-544193625). Could it be that somehow the issue is not surfacing on 3.8 / Uvicorn 0.10+?

I saw your comment, and I didn't understand why this would cause an exception at all.  I assumed you were doing something else, like killing the server at the right moment, as mentioned in earlier comments by others.

> 
>     * A `premature_close`-type endpoint does yield the exact same `RemoteProtocolError` than the one listed in #96:
> 
> 
> ```python
> >>> import httpx
> >>>httpx.get("http://localhost:8000/premature_close")
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/api.py", line 125, in get
>     return request(
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/api.py", line 84, in request
>     return client.request(
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/client.py", line 845, in request
>     return self.send(
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/client.py", line 908, in send
>     response.read()
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/models.py", line 1097, in read
>     self._content = b"".join([part for part in self.stream()])
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/models.py", line 1097, in <listcomp>
>     self._content = b"".join([part for part in self.stream()])
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/models.py", line 1108, in stream
>     for chunk in self.raw():
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/models.py", line 1136, in raw
>     for part in self._raw_stream:
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/concurrency/base.py", line 159, in iterate
>     yield self.run(async_iterator.__anext__)
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/concurrency/asyncio.py", line 260, in run
>     return self.loop.run_until_complete(coroutine(*args, **kwargs))
>   File "/Users/florimond/.pyenv/versions/3.8.0/lib/python3.8/asyncio/base_events.py", line 608, in run_until_complete
>     return future.result()
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/dispatch/http11.py", line 152, in _receive_response_data
>     event = await self._receive_event(timeout)
>   File "/Users/florimond/Developer/python-projects/httpx/httpx/dispatch/http11.py", line 164, in _receive_event
>     event = self.h11_state.next_event()
>   File "/Users/florimond/Developer/python-projects/httpx/venv/lib/python3.8/site-packages/h11/_connection.py", line 420, in next_event
>     event = self._extract_next_receive_event()
>   File "/Users/florimond/Developer/python-projects/httpx/venv/lib/python3.8/site-packages/h11/_connection.py", line 369, in _extract_next_receive_event
>     event = self._reader.read_eof()
>   File "/Users/florimond/Developer/python-projects/httpx/venv/lib/python3.8/site-packages/h11/_readers.py", line 169, in read_eof
>     raise RemoteProtocolError(
> h11._util.RemoteProtocolError: peer closed connection without sending complete message body (incomplete chunked read)
> ```
> 
> This error would be caught and handled by this PR though (it raises a `ConnectionClosed` error), so I think we're on a good track anyway. I haven't followed the whole story though so I'll let others make a more informed call. :)

File: tests/conftest.py (hunk 2 above)
Comment (toppk) [reply tests/conftest.py]: we probably could.  I simply wanted to avoid any chance of a race condition.

**Comment by toppk** [comment]
address some issues with original pr.

* renamed the test case
* learned about url.copy_with(path="")
* removed the sleep from the "bad" app.

Other then not really understanding how @florimondmanca test case ever triggered, I'm confident this PR addresses an uncaught exception issue.

**Comment by toppk** [comment]
pinging @tomchristie on this PR.   it would be nice to get this trivial patch merged before it becomes stale.

**Comment by florimondmanca** [comment]
So, I just tried this out locally and it works fab. Also handles the case when a server is shutdown while we're receiving the response. So, merging! Thanks @toppk

**Pull request merged**